Adding LiveData support into Databinding

wellingtoncabral opened the pull request:
As we talked, follows PR that uses Databinding with LiveData. This way, Activity becomes cleaner and only responds/observes to events that ViewModel should not know about.
Feel free to ask me about anything.
Thank you

wellingtoncabral committed:
Adding LiveData support into Databinding

wellingtoncabral committed:
Performed lint and ktlint

nglauber commented on presentation/src/main/java/dominando/android/presentation/BookListViewModel.kt:
```diff
@@ -22,19 +22,13 @@ import kotlinx.coroutines.withContext
-        return removeOperation
-    }
+    fun state(): LiveData<ViewState<List<BookBinding>>> = state
+    fun removeOperation(): LiveData<LiveEvent<ViewState<Unit>>> = removeOperation
```
`LiveEvent` is a `LiveData`. Why do we need do this wrapping?

wellingtoncabral replied:
`LiveEvent` is not a `LiveData`. It is a simple generic class that allows reading the content only once, right?

nglauber replied:
Oh! you're right! 😅

nglauber commented on presentation/src/main/java/dominando/android/presentation/BookFormViewModel.kt:
```diff
@@ -21,34 +24,62 @@ class BookFormViewModel(
             try {
                 withContext(Dispatchers.IO) {
-                    saveBookUseCase.execute(BookConverter.toData(book))
+                    book.value?.let { saveBookUseCase.execute(BookConverter.toData(it)) }
```
If `book` is null, it will do a false successful save operation. We must must first validate if the book is not null.

wellingtoncabral replied:
Fixed up

nglauber commented on app/src/main/java/dominando/android/livros/BookDetailsFragment.kt:
```diff
@@ -8,51 +8,38 @@ import android.view.MenuItem
 
-    private lateinit var binding: FragmentBookDetailsBinding
+    private val viewModel: BookDetailsViewModel by inject {
+        parametersOf(arguments?.getParcelable<Book>("book")?.id)
```
I know it's from existing code, but could you create a constant for the argument key? e.g.: `EXTRA_BOOK`

wellingtoncabral replied:
I created a Constants class to put all public constants

nglauber commented on app/src/main/java/dominando/android/livros/BookFormFragment.kt:
```diff
@@ -33,78 +31,53 @@ class BookFormFragment : BaseFragment() {
         super.onViewCreated(view, savedInstanceState)
-        if (viewModel.book == null) {
-            viewModel.book = arguments?.getParcelable("book") ?: Book()
+        arguments?.getParcelable<Book>("book")?.let {
```
Use `EXTRA_BOOK` constant here. (see comment above)

wellingtoncabral replied:
The same solution above

nglauber commented on app/src/main/java/dominando/android/livros/BookFormFragment.kt:
```diff
@@ -33,78 +31,53 @@ class BookFormFragment : BaseFragment() {
     private fun init() {
-        viewModel.getState().observe(viewLifecycleOwner, Observer { event ->
-            event?.peekContent()?.let { state ->
+        viewModel.state.observe(viewLifecycleOwner, Observer { event ->
```
Running the app, I noticed that a message is displayed if the user try to save a book with empty fields.
If the device is rotated, the message is displayed again.

wellingtoncabral replied:
I have seen this bug.
To solve it I needed to create the `SingleLiveEvent` class. This is Google's solution to the one-shot event using LiveData. 
https://github.com/android/architecture-samples/blob/dev-todo-mvvm-live/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/SingleLiveEvent.java

So we don't need to use `LiveEvent` class anymore

nglauber commented on app/src/main/java/dominando/android/livros/BookListFragment.kt:
```diff
@@ -89,16 +103,15 @@ class BookListFragment : BaseFragment() {
-                router.showBookDetails(book)
-            }
-            attachSwipeToRecyclerView()
+    private fun initRecyclerView(recyclerView: RecyclerView) {
```
Why pass the `RecyclerView` as parameter once it is a member field?

wellingtoncabral replied:
Fixed up

nglauber commented on presentation/src/main/java/dominando/android/presentation/BookDetailsViewModel.kt:
```diff
@@ -15,38 +15,36 @@ import kotlinx.coroutines.flow.flowOn
+    val book = Transformations.map(state) { it.data }
 
-    fun getState(): LiveData<ViewState<BookBinding>> = state
+    init { loadBook() }
```
Could we use a `@LifecycleEvent` instead?

wellingtoncabral replied:
Using `@LifecycleEvent` instead of `init block`, it's necessary to check if the `book` variable is already loaded every time when the device is rotated to avoid a new repository request.
Using `init block` approach, the `book` variable will be loaded only once when the ViewModel is created. When the device is rotated ViewModel will return the retained value without extra checking.
What do you think?

nglauber commented on app/src/main/java/dominando/android/livros/BookListFragment.kt:
```diff
@@ -89,16 +103,15 @@ class BookListFragment : BaseFragment() {
+        attachSwipeToRecyclerView()
+    }
+
+    private fun initFab(fab: FloatingActionButton) {
```
Same here. Why pass the `FAB` as parameter once it is a member field?

wellingtoncabral replied:
Fixed up

nglauber reviewed (changes requested):
Thanks for your contribution @wellingtoncabral! I really appreciated your PR. There are a bunch of interesting stuff in here. Could you please check my comments?
Also, the unit tests for the presentation layer has compilation errors :(

wellingtoncabral committed:
- Check if book variable is null before save it
- Create a Constants class to put all public constants. e.g. EXTRA_BOOK
- Added SingleLiveEvent class. This is Google's solution to the one-shot event using LiveData
- Removed LiveEvent class
